wip-stringparsexml.md

```diff
@@ -0,0 +1,174 @@
+<pre>
+  WIP: WIP-00XX
+  Layer: Consensus (hard fork)
+  Title: Add StringParseXMLMap to RADON operators
+  Authors: Luis Rubio <luisrwitnet.foundation>
+  Discussions-To: `#dev-general` channel on Witnet Community's Discord server
+  Status: Draft
+  Type: Standards Track
+  Created: 2021-12-14
+  License: BSD-2-Clause
+</pre>
+
+
+## Abstract
+
+The **StringParseXMLMap** is proposed to be included in the operators catalog available in the RAD engine. Data requests
+will then support XML parsing as part of the **retrieval** scripts.
+
+
+## Motivation and rationale
+
+In order to increase the flexibility and potential use cases of the Witnet protocol, it is paramount to continue 
+expanding the catalog of available operators.
+
+This document proposes adding support for **XML parsing** allowing the protocol to extract information from XML APIs
+or XML-encapsulated strings to be retrieved in Witnet by witnesses
+
+## Specification
+
+The *StringParseXMLMap* operator converts a *RadonString* (where the XML is defined) in a *RadonMap* following the next
+rules:
+- Every element will be defined as a *RadonMap* where the **tag** information would be inserted with the *_label* key, 
+  **attributes** would use the *@* symbol to precede their key name, and the elements inside would be added in a 
+  *RadonArray* with the key *_children*.
+  
+```xml
+<tag1 attr1=001>
+    <in_tag1>[...]</in_tag1>
+    <in_tag2>[...]</in_tag2>
+</tag1>
+```
+```rust
+{
+    _label: "tag1"
+    @attr1: "001"
+    _children: [
+        {in_tag1[...]}, 
+        {in_tag2[...]},
+    ]
+}
+```
+
+- Elements with text content that have no attributes or more elements inside (aka text children) would be added as a 
+  part of the map, using the *tag* as a key and the *content* as a value
+
+```xml
+<tag1 attr1=001>
+    <in_tag1>[...]</in_tag1>
+    <in_tag2>[...]</in_tag2>
+    <in_tag3>Hello</in_tag3>
+</tag1>
+```
+```rust
+{
+    _label: "tag1"
+    @attr1: "001"
+    in_tag3: "Hello"
+    _children: [
+        {in_tag1[...]}, 
+        {in_tag2[...]},
+    ]
+}
+```
+- In the case of elements that have several elements, some of them with text content, or text content with attributes.
+  Those text elements would be included directly in the map with the key *_test#n* where *n* is the text counter
```

Comment:
```suggestion
- Elements that have several children with text content and with or without attributes MUST include their children directly into the *Map* that represents them using the key `_text#n` where `n` is the indexed position of each of the children elements.
```

Reply:
Now I wonder... what's the RADON representation of this perfectly valid XML document? :thinking: 
```xml
<tag1 attr1=001>
    <in_tag1>
        Hello
        <in_in_tag1>World</in_in_tag1>
        Witnet
        <in_in_tag1>GM</in_in_tag1>
        frens
    </in_tag1>
    <in_tag2 attr="hello">World</in_tag2>
</tag1>
```

Reply:
Error because of missing `xmlns`. And error because `attr1=001` is invalid, must be `attr1="001"`, at least according to the library we are using. But ignoring that, this is the output:

```
RadonMap {
    value: {
        "@attr1": String(RadonString {
            value: "001"
        }),
        "_children": Array(RadonArray {
            value: [Map(RadonMap {
                value: {
                    "_label": String(RadonString {
                        value: "in_tag1"
                    }),
                    "_text#0": String(RadonString {
                        value: "Hello"
                    }),
                    "_text#1": String(RadonString {
                        value: "Witnet"
                    }),
                    "_text#2": String(RadonString {
                        value: "frens"
                    }),
                    "in_in_tag1": String(RadonString {
                        value: "GM"
                    })
                }
            }), Map(RadonMap {
                value: {
                    "@attr": String(RadonString {
                        value: "hello"
                    }),
                    "_label": String(RadonString {
                        value: "in_tag2"
                    }),
                    "_text#0": String(RadonString {
                        value: "World"
                    })
                }
            })],
            is_homogeneous: true
        }),
        "_label": String(RadonString {
            value: "tag1"
        }),
        "_xmlns": String(RadonString {
            value: "https://witnet.io/"
        })
    }
}
```

Reply:
O_O, we lost the first in_in_tag1